### ISSUE TITLE
Row validation happens before undo stack is updated on updateFromArray

### ISSUE THREAD

**Issue by ottosson** [handsontable/handsontable#10297]
### Description
I want to add rows with data to my datagrid without the user being able to undo these actions. So I make a copy of the undo stack before I add rows and data, then replace the stack containing "add rows" and "insert data" with the old stack. This works quite well, but if I have cells that needs validating, like dropdown cells, the validation seems to comes before the undo stack is updated, causing me to replace the stack before all actions have been added to the stack. So now the user can undo parts of the row adding. I can't understand this really. It feels like the validation is done asynchronously and then the action is added to undo stack after it's done validating somehow.

I would expect the undo stack to get updated before any validation occurs, or that the validation would happen synchronously so that when I add rows and data the stack should contain all actions when I try to remove them.

Maybe I'm doing something wrong or there is a more proper way to suspend things from being added to the undo stack.

### Demo
This gets really noticable if the table has to add rows to be able to insert the data: https://jsfiddle.net/mhsg6b81/
Check the console for output.

**Comment by AMBudnik** [comment]
Hi @ottosson 

That is correct, we can see here https://jsfiddle.net/4n2ct5fg/3/ that the `beforeValidate` calls are made before we get the calls from the `beforeUndoStackChange` hook. 

The hook that runs before the `beforeValidate` is `beforeChange` https://jsfiddle.net/rwbk8s47/. So between `beforeChange` and `afterChange`, we get the difference in actions https://jsfiddle.net/965yju2g/.

**Comment by ottosson** [comment]
Thanks! That works

**Comment by ottosson** [comment]
Sorry, I was too quick to celebrate. Seems to me that you still can undo the actions made in the batch.

**Comment by ottosson** [comment]
Thanks @AMBudnik, but this didn't help unfortunately. You can still undo events after the rows have been added...

Not sure if I understand, but doesn't this feel like unintended behavior? Can you explain why this feels asynchronous?

1. Start of addRows method
2. Store copy of undo stack in variable
3. Rows are added
4. Event pushed to undo stack
5. Rows are populated with data (no undo yet)
6. End of addRows, we reset the undo stack
7. Validation happens
8. Undo stack is pushed with the add rows event
9. Now we have an undo stack with populate data event

**Comment by AMBudnik** [comment]
Hi @ottosson 

I am coming with an update where we can do some changes before we add the rows, then click the button and be able to undo only the changes that do not related to the button action. I just executed the `pop()` on the array of `doneActions` x operations added by the method. Here's a demo https://jsfiddle.net/sakvxz1f/.

**Comment by ottosson** [comment]
Hm, using setTimeout to wait for an arbitrary time and then pop the actions feels hacky. How do you really know it's done after 1000ms. Maybe we are working on a slow machine and adding lots of rows. Maybe then 1000ms is not enough.
Also if you use setTimeout you could just as well just reser the undoActions to the `doneActionsBeforeAdd` variable: https://jsfiddle.net/17wdzgco/

**Comment by AMBudnik** [comment]
> Maybe then 1000ms is not enough.

That is true. You can't be sure. I already raised the question in our internal chat and will update you as soon as I get any response.

**Comment by AMBudnik** [comment]
Sorry that it takes so much time @ottosson 
I did not forget about this issue. I am still waiting for the feedback.

**Comment by ottosson** [comment]
Anything new here?

**Comment by AMBudnik** [comment]
Although we experienced some delays due to holidays and holiday breaks, the research is still scheduled to be completed, and we are actively working towards making progress on the development issue.

**Comment by ottosson** [comment]
Good to know, thanks! I'll just try to be patient then :P

**Comment by ottosson** [comment]
Ok, thanks! I still really hope you'll get to this issue soon. If there are no work-arounds I'd consider this a bug. And it's affecting me quite much in specific cases.

**Comment by AMBudnik** [comment]
Hi @ottosson 

I have updates from our dev team.

What about setting up the condition based on the `actionType`? Here's a demo for tests https://jsfiddle.net/e6p14qdf/2/ it may need some adjustments, but I think that it's working trying to give it a go.

**Comment by ottosson** [comment]
Hi. Thanks for the suggestion. But I'm not sure this quite works unfortunately...

In my case I do both `hot.alter("insert_row_below", ...)` and `hot.populateFromArray(... data)` and want to clear them both when the specific add function runs but not when other add functions run. If I add data I will have to remove the "populate" event as well, and add logic to check if there are both a insert and populate... Feels really hacky and unreliable.

Shouldn't the undo stack be updated directly after the event? Don't know the internals of HoT but would it be difficult to change?

**Comment by AMBudnik** [comment]
Sorry, I missed the notification earlier, @ottosson. I assume the issue has been resolved by now, but if this is still a problem, please let me know. We can also continue the conversation over email if that is more convenient. I am closing the issue for now, but feel free to reach out anytime.
